src/extension.ts

```diff
@@ -383,22 +367,70 @@ async function terraformCommand(command: string, languageServerExec = true): Pro
 }
 
 function enabled(): boolean {
-  return config('terraform').get('languageServer.external', false);
+  return config('terraform').get('languageServer.enable', false);
 }
 
-async function migrateLegacySettings() {
-  if (config('terraform').has('languageServer.enabled')) {
-    try {
-      await config('terraform').update(
-        'languageServer',
-        { enabled: undefined, external: true },
-        vscode.ConfigurationTarget.Global,
-      );
-    } catch (err) {
-      const error = err instanceof Error ? err.message : err;
-      console.error(`Error trying to erase pre-2.0.0 settings: ${error}`);
-    }
+async function migrateLegacySettings(ctx: vscode.ExtensionContext) {
+  ctx.globalState.update('terraform.disableSettingsMigration', false);
+  if (ctx.globalState.get('terraform.disableSettingsMigration', false)) {
+    return;
+  }
+
+  const warnMigration = await warnIfMigrate([
+    'terraform.languageServer.external',
+    'terraform.languageServer.pathToBinary',
+    'terraform-ls.rootModules',
+    'terraform-ls.excludeRootModules',
+    'terraform-ls.ignoreDirectoryNames',
+    'terraform-ls.terraformExecPath',
+    'terraform-ls.terraformExecTimeout',
+    'terraform-ls.terraformLogFilePath',
+    'terraform-ls.experimentalFeatures',
+  ]);
+  if (warnMigration === false) {
+    return;
   }
+
+  const choice = await vscode.window.showInformationMessage(
+    'Terraform configuration settings have changed in the latest update',
+    {
+      detail: 'You can automatically migrate or ask to not be reminded again and migrate the settings yourself',
+      modal: true,
+    },
+    { title: 'Migrate' },
+    { title: "Don't remind me again" },
+  );
+  if (choice === undefined) {
+    return;
+  }
+
+  switch (choice.title) {
+    case "Don't remind me again":
+      ctx.globalState.update('terraform.disableSettingsMigration', true);
+      return;
+    case 'Migrate':
+    // migrate below
+  }
+
+  await migrate('terraform', 'languageServer.external', 'languageServer.enable');
+  await migrate('terraform', 'languageServer.pathToBinary', 'languageServer.path');
+  await migrate('terraform', 'languageServer.args', 'languageServer.args');
+  await migrate('terraform', 'languageServer.ignoreSingleFileWarning', 'languageServer.ignoreSingleFileWarning');
```

Comment:
The `migrate` function detects if the setting is present at any of the scopes we support, then writes it out with the current value. If it's not defined, then it does nothing and returns.

Previously `terraform.languageServer` was written as a JSON object. We are extracting this out into separate settings. If a user had the block written out with any or all of the settings defined, we have to migrate it otherwise the user is left with different settings in different formats. The older object based notation will be grayed out and inactivate, but still present with the new format side by side.

For example, a user used autocomplete in the JSON file and had the following populated thanks to our `default` stanza on line 313:

```
"terraform.languageServer": {
  "external": true,
  "pathToBinary": "",
  "args": [
    "serve"
  ],
  "ignoreSingleFileWarning": false
}
```

They then edited the file like so (with made up argument values):

```
"terraform.languageServer": {
  "external": true,
  "args": ["serve", "--foo"],
}
```

If we did not account for `args`, ran the migration, then the settings would look like:

```
"terraform.languageServer": {
  "external": true,
  "args": ["serve", "--foo"],
}
"terraform.languageServer.enable": true,
```

Our extension would pick up `terraform.languageServer.enable` correctly, but miss `terraform.languageServer.args` and the user would not know why. The terraform-ls binary would be started with 'server' and miss '--foo'.